py_ecc/field_elements.py

```diff
@@ -40,21 +23,31 @@ def inv(a: int, n: int) -> int:
 # and it becomes a field element.
 class FQ(object):
     n = None  # type: int
+    field_modulus = None
+    curve_name = None
+
+    def __init__(self, val: IntOrFQ, curve_name: str) -> None:
+        """
+        curve_name can be either 'bn128' or 'bls12_381'
+        This is needed to obtain field_modulus, FQ2_MODULUS_COEFFS
+        and FQ12_MODULUS_COEFFS from the curve properties
+        """
+        self.curve_name = curve_name
```

Comment:
My *gut* is telling me that this `curve_name` based approach could be improved.  The primary downside that I want to get rid of is how much we seem to have to pass this parameter around.  If this was instead a property of the class....

```python
class FQ(ABC):
    field_modulus = None

    def __init__(self, val):
        if self.field_modulus is None:
           raise SomeError("...")
```

This sets a simple baseline for ensuring that `FQ` classes and their derivatives don't get instantiated without their `field_modulus`

```python
class FQ_BN128(FQ):
    field_modulus = field_properties["bn128"]["field_modulus"]
```

Now we have an `FQ_BN128` which represents the `FQ` for the bn128 curve.

```python
class FQ:
    ...

    def __add__(self, other):
        on = other.n if isinstance(other, FQ) else other
        return type(self)(self.n + on) % self.field_modulus)

    @classmethod
    def one(cls):
        return cls(1)
```

This allows us to remove the passing around of the `self.curve_name` within the `FQ` object methods quite easily.

I did a quick search and this seems to cleanly remove the need to pass the curve name around.  What do you think?

Reply:
@pipermerriam I agree with this, but I was left with no other option. I have thought of this approach, but it would involve a lot of classes in the code. One class should be declared for each of the curves `bn128`, `bls12_381`, `optimized_bn128`, `optimized_bls12_381` curve ; that too for only `FQ`. For `FQP`, `FQ2` and `FQ12`, the same process needs to be repeated, and we would be ending up with `16 classes` in total. Hence I just resented to this approach.

I know this design of using `curve_name` leads to a lot parameter passing, but I felt it was the best choice, given the above reasons. Please correct me if wrong.